### PR TITLE
os/bluestore/bluestore_tool: compare retval stat() with -1

### PR DESCRIPTION
**Pull request by tchaikov**
Fixes: https://tracker.ceph.com/issues/50891

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by ifed01 — APPROVED** [review]

**Comment by tchaikov** [comment]
changelog

- add a space after `if`.

File: src/os/bluestore/bluestore_tool.cc [hunk 1]
```diff
@@ -780,45 +778,52 @@ int main(int argc, char **argv)
...
-      exit(EXIT_FAILURE);
     }
 
+    auto [target_path, has_size_spec] =
+      [&dev_target]() -> std::pair<string, bool> {
+      if (dev_target.empty()) {
+	return {"", false};
+      }
+      std::error_code ec;
+      fs::path target_path = fs::weakly_canonical(dev_target, ec);
```
Comment (tchaikov) [inline src/os/bluestore/bluestore_tool.cc:789]:
```
-- The CXX compiler identification is GNU 7.5.0
-- The C compiler identification is GNU 7.5.0
...
/build/ceph-17.0.0-4436-g0a8ba2e2/src/os/bluestore/bluestore_tool.cc:789:34: error: 'weakly_canonical' is not a member of 'fs'
       fs::path target_path = fs::weakly_canonical(dev_target, ec);
                                  ^~~~~~~~~~~~~~~~
```

turns out libstdc++ shipped along with GCC 7.5.0 does not offer `std::filesystem::weakly_canonical()` yet. and we cannot drop bionic yet. see https://github.com/ceph/ceph-build/pull/1794#issuecomment-815435506

**Comment by tchaikov** [comment]
changelog

- add a commit to use `boost::filesystem` as an alternative of `std::filesystem`

**Comment by ifed01** [comment]
@tchaikov - wouldn't it be better at this point to split this PR into two parts:
1) proper stat() retval handling
2) refactoring to start using filesystem

The rationale is that part 1) is definitely worth prompt backporting along with my ceph-volume migrate feature implementation.
Part 2) can keep living in master for a while (or even be not backported at all as this is rather a code cleanup)

What do you think?

**Comment by tchaikov** [comment]
@ifed01 but i do split the PR into separated commits.

**Comment by ifed01** [comment]
> @ifed01 but i do split the PR into separated commits.

OK, nevermind

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-21_12:49:52-rados-wip-kefu-testing-2021-05-21-1851-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-22_15:42:09-rados-wip-kefu-testing-2021-05-22-2207-distro-basic-smithi/

failure tracked by 

- https://tracker.ceph.com/issues/50243
- https://tracker.ceph.com/issues/50151

**Pull request merged**